File: NMS/v1_18_R2/src/main/java/xyz/oribuin/beepet/nms/v1_18_R2/NMSHandlerImpl.java
package xyz.oribuin.beepet.nms.v1_18_R2;

import org.bukkit.Location;
import org.bukkit.NamespacedKey;
import org.bukkit.craftbukkit.v1_18_R2.CraftWorld;
import org.bukkit.entity.Bee;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.persistence.PersistentDataType;
import xyz.oribuin.beepet.nms.CustomBeeSettings;
import xyz.oribuin.beepet.nms.NMSHandler;

import java.util.Random;
import java.util.function.Consumer;

import static xyz.oribuin.beepet.BeePetPlugin.OWNER_KEY;

public class NMSHandlerImpl implements NMSHandler {

    @Override
    public void spawn(Player owner, Location location, CustomBeeSettings options, Consumer<Bee> beeConsumer) {
        ServerBee serverBee = new ServerBee(owner, options);
        Location loc = location.clone();
        if (loc.getWorld() == null) return;

        boolean spawned = ((CraftWorld) loc.getWorld()).getHandle().addFreshEntity(serverBee);
        if (!spawned || OWNER_KEY == null) return;


        Bee bukkitEntity = (Bee) serverBee.getBukkitEntity();
        bukkitEntity.teleport(location); // Teleport the bee to the location
        bukkitEntity.getPersistentDataContainer().set(
                OWNER_KEY,
                PersistentDataType.STRING,
                owner.getUniqueId().toString()
        );

        bukkitEntity.setCustomName(options.getName());
        bukkitEntity.setCustomNameVisible(true);
        beeConsumer.accept(bukkitEntity);
    }

    @Override
    public void remove(Entity entity) {
        entity.remove(); // Remove the bee from the world
    }

}

File: NMS/v1_16_R1/src/main/kotlin/xyz/oribuin/beepet/nms/v1_16_R1/NMSHandlerImpl.kt
package xyz.oribuin.beepet.nms.v1_16_R1

import io.github.bananapuncher714.nbteditor.NBTEditor
import net.minecraft.server.v1_16_R1.NBTTagCompound
import org.bukkit.craftbukkit.v1_16_R1.CraftWorld
import org.bukkit.entity.Bee
import org.bukkit.entity.Player
import org.bukkit.event.entity.CreatureSpawnEvent
import org.bukkit.metadata.FixedMetadataValue
import org.bukkit.plugin.Plugin
import xyz.oribuin.skyblock.nms.BeeOptions
import xyz.oribuin.skyblock.nms.CustomOptions
import xyz.oribuin.skyblock.nms.NMSHandler

class NMSHandlerImpl : NMSHandler {

    override fun createBee(plugin: Plugin, owner: Player, options: BeeOptions, custom: CustomOptions): Bee {

        // Create bee instance
        val bee = CustomBee(owner, options, custom.sitting)
        val loc = owner.location.clone()

        // Spawn the bee into the world
        (owner.world as CraftWorld).handle.addEntity(bee, CreatureSpawnEvent.SpawnReason.CUSTOM)
        bee.setLocation(loc.x, loc.y, loc.z, loc.yaw, loc.pitch)

        // Add the correct values and meta data
        var bukkit = bee.bukkitEntity as Bee
        bukkit = NBTEditor.set(bukkit, owner.uniqueId.toString(), "beepet")
        bukkit.setMetadata("beepet", FixedMetadataValue(plugin, owner.uniqueId.toString()))

        bukkit.isCustomNameVisible = true
        bukkit.customName = custom.name

        return bukkit
    }

}